tests/upgrades/test_hostgroup.py

```diff
@@ -16,155 +16,166 @@
+
+    :expectedresults: After upgrade
+        1- Hostgroup entities remain same.
+        2- Hostgroup entties update should work.
+        3- Hostgroup cloned should work.
+        4- Cloned hostgroup should be the subset of original hostgroup.
+        5- Hostgroup entities deletion should work
+
+    """
+    pre_test_name = [
+        mark.kwargs['depend_on'].__name__
+        for mark in request.node.own_markers
+        if 'depend_on' in mark.kwargs
+    ][0]
+
+    # verify host-group is intact after upgrade
+    org = entities.Organization().search(query={'search': f'name="{pre_test_name}_org"'})[0]
+    request.addfinalizer(org.delete)
+    loc = entities.Location().search(query={'search': f'name="{pre_test_name}_loc"'})[0]
+    request.addfinalizer(loc.delete)
+    proxy = entities.SmartProxy().search(
+        query={'search': f'url = https://{settings.server.hostname}:9090'}
+    )[0]
+    hostgrp = entities.HostGroup().search(query={'search': f'name={pre_test_name}_host_grp'})[0]
+    request.addfinalizer(hostgrp.parent.delete)
+    request.addfinalizer(hostgrp.delete)
+    assert pre_test_name + "_host_grp" == hostgrp.name
+    assert proxy.id == hostgrp.puppet_proxy.id
+    assert proxy.id == hostgrp.puppet_ca_proxy.id
+
+    domain = entities.Domain().search(query={'search': f'name={pre_test_name}_domain'})[0]
+    assert domain.id == hostgrp.domain.id
+    request.addfinalizer(domain.delete)
+
+    subnet = entities.Subnet().search(query={'search': f'name={pre_test_name}_subnet'})[0]
+    assert subnet.id == hostgrp.subnet.id
+    request.addfinalizer(subnet.delete)
+
+    parent = entities.HostGroup().search(
+        query={'search': f'name={pre_test_name}_parent_host_grp'}
+    )[0]
+    assert parent.id == hostgrp.parent.id
+
+    os = entities.OperatingSystem().search(query={'search': f'name={pre_test_name}_os'})[0]
+    assert os.id == hostgrp.operatingsystem.id
+    request.addfinalizer(os.delete)
+
+    # update host-group after upgrade
+    new_name = gen_string('alpha')
+    hostgrp.name = new_name
+    hostgrp.update(['name'])
+    assert new_name == hostgrp.name
+
+    new_subnet = entities.Subnet().create()
+    hostgrp.subnet = new_subnet
+    hostgrp.update(['subnet'])
+    assert new_subnet.id == hostgrp.subnet.id
+
+    new_domain = entities.Domain().create()
+    hostgrp.domain = new_domain
+    hostgrp.update(['domain'])
+    assert new_domain.id == hostgrp.domain.id
+
+    new_os = entities.OperatingSystem().create()
+    hostgrp.operatingsystem = new_os
+    hostgrp.update(['operatingsystem'])
+    assert new_os.id == hostgrp.operatingsystem.id
+
+    # clone hostgroup
+    hostgroup_cloned_name = gen_string('alpha')
+    hostgroup_cloned = entities.HostGroup(id=hostgrp.id).clone(
+        data={'name': hostgroup_cloned_name}
+    )
+
+    hostgroup_cloned_object = entities.HostGroup().search(
+        query={'search': f'name={hostgroup_cloned_name}'}
+    )[0]
```

Comment:
No assertions made for cloned host-group, add one below to check if the cloned host group exists.

Or modify above as :

```suggestion
   with not pytest.asertRaises(HTTPError):
        hostgroup_cloned_object = entities.HostGroup().search(
            query={'search': f'name={hostgroup_cloned_name}'}
        )[0]
```

I am not exactly sure if its HTTP Error, just check and add appropriatly.

Reply:
No need to put the check here, if the cloned host group does not exist then the test case will get failed with "IndexError" 
And here HTTPError does not work because the search module returns an empty list if no match found.

Reply:
Not blocking, but a good pattern to follow here is to actually capture expected failure modes and call `pytest.fail`, instead of letting the test result in an `Error` state instead of a failed state. Its far more clear that the test step failed because of something predictable and implicit, and didn't 'error' out within the framework. 

There's an implicit assertion here that the result of `entities.HostGroup().search()` contains a result - and an assumption that it only contained the one we were searching for.

A better pattern could look like the following, where we first explicitly assert that only one result was returned, and then only index against that search result. If there is a failure, it will be an assertion exception - which pytest handles as a failure instead of an error.

```
hostgroup_search = entities.HostGroup().search(search': f'name={hostgroup_cloned_name}'})
assert len(hostgroup_search) == 1
hostgroup_cloned_object = hostgroup_search[0]
```

Reply:
I think pytest also treats IndexError as a failure than just an error, but assert is more appropriate here.

```
>       org = entities.Organization().search(query={'search': f'name="{pre_test_name}_org"'})[0]
E       IndexError: list index out of range

tests/upgrades/test_hostgroup.py:103: IndexError
================================================================================================ warnings summary =================================================================================================
FAILED tests/upgrades/test_hostgroup.py::test_post_crud_hostgroup - IndexError: list index out of range
=================================================================================== 1 failed, 1 deselected, 7 warnings in 1.36s ===================================================================================


        org_search = entities.Organization().search(query={'search': f'name="{pre_test_name}_org"'})
>       assert len(org_search) == 1
E       assert 0 == 1
E        +  where 0 = len([])

tests/upgrades/test_hostgroup.py:104: AssertionError
================================================================================================ warnings summary ============================================================================================================================================================================================= short test summary info =============================================================================================
FAILED tests/upgrades/test_hostgroup.py::test_post_crud_hostgroup - assert 0 == 1
=================================================================================== 1 failed, 1 deselected, 7 warnings in 1.28s ===================================================================================


```